text/0000-async.md

```diff
@@ -0,0 +1,916 @@
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
+
+### Task contexts
+
+All tasks are executed with two pieces of contextual information:
+
+- A `Waker` for waking the task up later on.
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    pub fn new(waker: &'a Waker, executor: &'a mut Executor) -> Context<'a>
+
+    /// Get the `Waker` associated with the current task.
+    pub fn waker(&self) -> &Waker;
+
+    /// Run an asynchronous computation to completion on the default executor.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the default executor is unable to spawn.
+    /// To handle executor errors, use the `executor` method instead.
+    pub fn spawn(&mut self, f: impl Future<Output = ()> + 'static + Send);
+
+    /// Get the default executor associated with this task.
+    ///
+    /// This method is useful primarily if you want to explicitly handle
+    /// spawn failures.
+    ///
+    /// NB: this will remain unstable until the final `Executor` trait is ready.
+    pub fn executor(&mut self) -> &mut BoxExecutor;
+}
+```
+
+Note that the `spawn` method here will box until `Executor` is added.
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Output;
+
+    /// Attempt to resolve the computation to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - `Poll::Pending` if the value is not ready yet.
+    /// - `Poll::Ready(val)` with the result `val` upon completion.
+    ///
+    /// Once a future has completed, clients should not `poll` it again.
+    ///
+    /// When a future is not ready yet, `poll` returns `Poll::Pending`.
+    /// The future will *also* register the interest of the current task in the
+    /// value being produced. For example, if the future represents the availability
+    /// of data on a socket, then the task is recorded so that when data arrives,
+    /// it is woken up (via `cx.waker()`). Once a task has been woken up,
+    /// it should attempt to `poll` the future again, which may or may not
+    /// produce a final value at that time.
```

Comment:
(reiterating comment that was lost in RFC update): As I said in the [last RFC](https://github.com/rust-lang/rfcs/pull/2395/files#r179835007), I think it's important to make it explicit that it's not an error to poll a future even before a registered 'interest' has caused a task wakeup (otherwise you can't be sure you can implement join/select).